numba/targets/linalg.py

```diff
@@ -2493,3 +2504,227 @@ def matrix_power_impl(a, n):
+
+# This is documented under linalg despite not being in the module
+
+
+@overload(np.trace)
+def matrix_trace_impl(a, offset=0):
+    """
+    Computes the trace of an array.
+    """
+
+    _check_linalg_matrix(a, "trace", la_prefix=False)
+
+    if not isinstance(offset, types.Integer):
+        raise TypeError("integer argument expected, got %s" % offset)
+
+    def matrix_trace_impl(a, offset=0):
+        rows, cols = a.shape
+        k = offset
+        if k < 0:
+            rows = rows + k
+        if k > 0:
+            cols = cols - k
+        n = max(min(rows, cols), 0)
+        ret = 0
+        if k >= 0:
+            for i in range(n):
+                ret += a[i, k + i]
+        else:
+            for i in range(n):
+                ret += a[i - k, i]
+        return ret
+
+    return matrix_trace_impl
+
+
+def _check_scalar_or_lt_2d_mat(a, func_name, la_prefix=True):
+    prefix = "np.linalg" if la_prefix else "np"
+    interp = (prefix, func_name)
+    # checks that a matrix is 1 or 2D
+    if isinstance(a, types.Array):
+        if not a.ndim <= 2:
+            raise TypingError("%s.%s() only supported on 1 and 2-D arrays "
+                              % interp)
+
+
+def _get_as_array(x):
+    if not isinstance(x, types.Array):
+        @register_jitable
+        def asarray(x):
+            return np.array((x,))
+        return asarray
+    else:
+        @register_jitable
+        def asarray(x):
+            return x
+        return asarray
+
+
+def _get_outer_impl(a, b, out):
+    a_arr = _get_as_array(a)
+    b_arr = _get_as_array(b)
+
+    if out in (None, types.none):
+        @register_jitable
+        def outer_impl(a, b, out):
+            aa = a_arr(a)
+            bb = b_arr(b)
+            return np.multiply(aa.ravel().reshape((aa.size, 1)),
+                               bb.ravel().reshape((1, bb.size)))
+        return outer_impl
+    else:
+        @register_jitable
+        def outer_impl(a, b, out):
+            aa = a_arr(a)
+            bb = b_arr(b)
+            np.multiply(aa.ravel().reshape((aa.size, 1)),
+                        bb.ravel().reshape((1, bb.size)),
+                        out)
+            return out
+        return outer_impl
+
+
+if numpy_version >= (1, 9):
+    @overload(np.outer)
+    def outer_impl(a, b, out=None):
+
+        _check_scalar_or_lt_2d_mat(a, "outer", la_prefix=False)
+        _check_scalar_or_lt_2d_mat(b, "outer", la_prefix=False)
+
+        impl = _get_outer_impl(a, b, out)
+
+        def outer_impl(a, b, out=None):
+            return impl(a, b, out)
+
+        return outer_impl
+else:
+    @overload(np.outer)
+    def outer_impl(a, b):
+
+        _check_scalar_or_lt_2d_mat(a, "outer", la_prefix=False)
+        _check_scalar_or_lt_2d_mat(b, "outer", la_prefix=False)
+
+        impl = _get_outer_impl(a, b, None)
+
+        def outer_impl(a, b):
+            return impl(a, b, None)
+
+        return outer_impl
+
+
+def _kron_normaliser_impl(x):
```

Comment:
Unfortunately, as with most linalg like functions, not trivially. I think if axis args can be supported in some more core functions then this could be written differently using `outer` and `concatenate` to support higher dimensions.